src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1347 @@
+          and returns it
+
+\*******************************************************************/
+
+symbol_exprt string_refine_preprocesst::new_symbol(
+  const std::string &prefix, const typet &type, bool static_lifetime)
+{
+  symbolt tmp_symbol;
+  // TODO: add a prefix to the name to avoid clashing with other
+  // variables of the program
+  std::ostringstream buf;
+  buf << prefix << "_" << (next_symbol_id++);
+  std::string name = buf.str();
+  tmp_symbol.base_name=name;
+  tmp_symbol.is_static_lifetime=static_lifetime;
+  tmp_symbol.is_state_var=static_lifetime;
+  tmp_symbol.is_type=false;
+  tmp_symbol.is_lvalue=static_lifetime; // not sure about that
+  tmp_symbol.mode=ID_java;
+  tmp_symbol.name=name;
+  tmp_symbol.type=type;
+  symbol_table.add(tmp_symbol);
+  return symbol_exprt(name,type);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::declare_function
+
+  Inputs: a name and a type
+
+ Purpose: declare a function with the given name and type
+
+\*******************************************************************/
+
+void string_refine_preprocesst::declare_function(
+  irep_idt function_name, const typet &type)
+{
+  auxiliary_symbolt func_symbol;
+  func_symbol.base_name=function_name;
+  func_symbol.is_static_lifetime=false;
+  func_symbol.mode=ID_java;
+  func_symbol.name=function_name;
+  func_symbol.type=type;
+  symbol_table.add(func_symbol);
+  goto_functions.function_map[function_name];
+}
+
+static bool is_java_string_type(const typet& type)
+{
+  if(type.id()==ID_symbol)
+  {
+    irep_idt tag=to_symbol_type(type).get_identifier();
+    return (tag==irep_idt("java::java.lang.String"));
+  }
+  else if(type.id()==ID_struct)
+  {
+    irep_idt tag=to_struct_type(type).get_tag();
+    return (tag==irep_idt("java.lang.String"));
+  }
+  else
+    return false;
+}
+
+static bool is_java_string_pointer_type(const typet& type)
+{
+  if(type.id()==ID_pointer)
+  {
+    pointer_typet pt=to_pointer_type(type);
+    typet subtype=pt.subtype();
+    return is_java_string_type(subtype);
+  }
+  else return false;
+}
+
+/*******************************************************************    \
+
+Function: string_refine_preprocesst::replace_string_pointer
+
+  Inputs: an expression
+
+ Outputs: an expression
+
+ Purpose: replace string pointers by expressions as expected by the string refinement;
+          should only be used in arguments of __CPROVER_uninterpreted functions.
+
+\*******************************************************************/
+
+exprt string_refine_preprocesst::replace_string_pointer(const exprt& in)
+{
+  if(is_java_string_pointer_type(in.type()))
+  {
+    exprt clsid=
+      symbol_exprt("java::java.lang.String", string_typet());
+    // TODO: we should use a function which is not currently in the code base
+    // to get the class identifier in this way:
+    // get_class_identifier_field(in, symbol_typet("java::java.lang.Object"), ns);
```

Comment:
changed this here https://github.com/diffblue/cbmc/pull/378/commits/918d97ad60c85ac5b956dd072dbdf70af057c7b5